### DISCUSSION TITLE
macOS Neovide: app only starts from the Terminal

### DISCUSSION THREAD

**Discussion by al3xandru** [neovide/neovide#2081] [Q&A]
I have installed Neovide on macOS Sonoma on a MBP M1. Starting the app by double-clicking doesn't do anything. But it does start when using `neovide` from the Terminal. 

I looked in Console for any error messages but I couldn't find much. Anyone knows why this is happening and how I could fix it?

thank you

**Comment by al3xandru** [comment]
I did set the path to nvim using `neovim_bin` setting but still nothing.

```
> cat ~/.config/neovide/config.toml
neovim_bin = "/opt/homebrew/bin/nvim"
frame = "Full"
```

**Comment by fredizzimo** [comment]
It's not ideal, but check this out https://github.com/neovide/neovide/issues/1859, there are several solutions there, but we need to come up with a better fix.

**Comment by al3xandru** [comment]
I read through the whole thread and the only hint is related to the system PATH.

On my system (macOS Sonoma on M1), I have installed nvim with brew. Thus `nvim` is in `/opt/homebrew/bin`. 

1. I have added `/opt/homebrew/bin` to `/etc/paths`. Restart. Nothing
2. I have symlinked nvim from `/usr/local/bin` which is in the PATH. Nothing.
3. The instruction of providing a path for neovim using `neovim_bin` in the toml config file doesn't work either.

Isn't the 3rd item a bug?  And unfortunately it's not clear what the solution is.

**Comment by al3xandru** [comment]
An update: I've learned that the issue of starting Neovide 0.11.2 in macOS might actually have nothing to do with the M1 or Intel versions. I have installed Neovide using brew on an Intel macOS running 13.5.2 and the behavior is identical. This make me believe that the path might be a red herring and the root cause is somewhere else (lack of permission? etc)

**Comment by fredizzimo** [comment]
There are a few of issues with workarounds
* https://github.com/neovide/neovide/issues/1373
* https://github.com/neovide/neovide/issues/1479
* https://github.com/neovide/neovide/issues/1998

**Comment by al3xandru** [comment]
From https://github.com/neovide/neovide/issues/1373:

1. $PATH should be set in `~/.zprofile`. I have create such a file and included export PATH=/opt/homebrew/bin:$PATH. Nothing. My system is set to use bash though.
2. `sudo launchctl config user path "$(brew --prefix)/bin:${PATH}` I tried this too. Nothing.
3. symlink nvim into /usr/local/bin. Tried. Nothing.
4. symlink the .app into $HOME/Applications. Nothing
5. "enable terminal inside Developer Tools" Not sure what this has to do. Tried. Nothing.

**Comment by al3xandru** [comment]
#1479 and #1998 seem to be about the app not being signed issue. I went through the dance of right clicking, etc.

**Comment by al3xandru** [comment]
I put together a quick AppleScript:
```
set theReturn to (do shell script "export PATH=/opt/homebrew/bin:/usr/local/bin:$PATH; echo $PATH; echo \"$(which nvim
)\"; open /Applications/Neovide.app")
log theReturn
```

Output:
```
/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
/opt/homebrew/bin/nvim
```

Next I changed it to:

```
set theReturn to (do shell script "export PATH=/opt/homebrew/bin:/usr/local/bin:$PATH; echo $PATH; echo \"$(which nvim
)\"; neovide")
log theReturn
```

The output this time:

```
error "ERROR: Unexpected output from neovim binary:
	/opt/homebrew/bin/nvim -v
Check that your shell doesn't output anything extra when running:
	$SHELL -lc '/opt/homebrew/bin/nvim -v'" number 1
```

It looks to me that Neovide it expecting a specific kind of output and it fails to parse it.

```
> nvim -v
NVIM v0.9.4
Build type: Release
LuaJIT 2.1.1696795921

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/opt/homebrew/Cellar/neovim/0.9.4/share/nvim"

Run :checkhealth for more info
```

**Reply by fredizzimo** [reply]
It looks like there is some extra output when you run it from your script, or when Neovide launches it, even if you don't see any in your test. Try this script:
```
set theReturn to (do shell script "export PATH=/opt/homebrew/bin:/usr/local/bin:$PATH; echo $PATH; echo \"$(which nvim
)\";$SHELL -lc '/opt/homebrew/bin/nvim -v'")
log theReturn
```

The cause might be the same as this issue https://github.com/neovide/neovide/issues/2049, something launches `ps` and because there's no proper terminal size it fails with some ouptut to the stderr. 

This message also contains a patch to get more output from Neovim https://github.com/neovide/neovide/issues/2049#issuecomment-1732667314, so if you are able to compile from source and apply that it would be helpful.

**Comment by al3xandru** [comment]
I have removed the other `echo` just to be sure. The output is:

```
NVIM v0.9.4
Build type: Release
LuaJIT 2.1.1696795921

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/opt/homebrew/Cellar/neovim/0.9.4/share/nvim"

Run :checkhealth for more info
```

**Reply by fredizzimo** [reply]
I guess we really need this patch https://github.com/neovide/neovide/issues/2049#issuecomment-1732667314 to debug further, and see what's actually output when Neovide runs `nvim --version`. It has to be something else than you see when it's run either manually or through the script modification that I made.

Note, with this change, the startup errors will be reported instead of just silently quitting, which would have made debugging it this far much easier.
* https://github.com/neovide/neovide/pull/2072

**Reply by al3xandru** [reply]
That seems to be a general improvement so why not committing it? I could then use brew to install the HEAD.

**Reply by fredizzimo** [reply]
I created a PR for it here https://github.com/neovide/neovide/pull/2083

**Comment by al3xandru** [comment]
I can confirm that after removing my .bash_profile, Neovide app could start. 

It has been impossible to figure out what’s breaking Neovide though. Even the first line of my bash_profile stty -ixon would already break it.

**Reply by fredizzimo** [reply]
Hm, shouldn't that be in `.bashrc`? Instead of `.bash_profile`? 

`.bashrc` is run for interactive shells, while `.bash_profile` is for noninteractive ones, so there might not be a tty available.

**Reply by al3xandru** [reply]
Maybe. 

Removing it though leaves another few hundred lines of bash and neovide still not starting :-(

**Comment by al3xandru** [comment]
The good news for me is that I got to a point where Neovide starts from the app on one of my machines.

The less good news is that I haven't been able to pinpoint the issue and that means others might hit the same problem and 1) there's no clear workaround for it; 2) there's no way to make neovide's start more resilient.

**Comment by fredizzimo** [comment]
I was wondering why you have so much stuff in your `.bash_profile` rather than `.bashrc`, but then I read this article https://scriptingosx.com/2017/04/about-bash_profile-and-bashrc-on-macos/.

IMO, the correct approach is to follow the advice there and add 
```bash
if [ -r ~/.bashrc ]; then
   source ~/.bashrc
fi
```
to your `.bash_profile`

And then at the top of the `.bashrc` the standard 
```bash
# If not running interactively, don't do anything
[[ $- != *i* ]] && return
```

This way you can separate the stuff you want to run on all shells, like setting paths to `.profile`, while adding everything you need for interactive shells in `.bashrc`, exactly like it works on Linux based systems.

This in itself won't help you pinpoint the issue, but if you only have paths (and perhaps aliases) in `.bash_profile`, there should not be anything preventing Neovide from starting. 

In `.bashrc` you can then have for example prompt customisation and other special stuff like `neofetch`

**Comment by fredizzimo** [comment]
@al3xandru if you are able to test then https://github.com/neovide/neovide/pull/2083, should include better error messages. It ill probably be merged to main soonish, after someone has reviewed it.

**Reply by al3xandru** [reply]
I have subscribed to #2083 to be notified when it gets merged and try it out. Is it that any echo command in the bash setup could mess with neovide's start?

**Reply by fredizzimo** [reply]
Yes, any output to either stdout or stderr will cause problems. 

The reason for that is that neovide use those to channels to communicate with neovim. That is a pretty standard way of doing things and for example most LSP clients communicate with neovim the same way.

What makes it problematic in the case of Neovide is that we also need to launch a new shell in case you run it from the desktop, so that all environment variables and aliases for example gets set up correctly.  And if the user has configured the shell incorrectly so that it outputs things also in a non-interactive mode then it will break. And worse if it waits for inputs, like oh-my-zsh update prompts for example

Perhaps we could somehow detect the initial input/output and ignore it, but things gets quite complex.
